### DISCUSSION TITLE
How to specify a runtime dependency like `org.tukaani:xz`

### DISCUSSION THREAD

**Discussion by IAmREGE** [FabricMC/fabric-api#4319] [Mod Dev Support] [answered]
### Description

In my `build.gradle.kts`, if I write:
```kotlin
dependencies {
    // ...
    modImplementation("org.tukaani:xz:1.10")
    // ...
}
```
The mod can run finely through `gradle runClient` and `gradle runServer`, but without Loom the game will crash with `java.lang.NoClassDefFoundError`.

**Comment by haykam821** [comment, marked as answer]
When you are running your mod in development, your dependencies will be available at runtime, but these dependencies are not available at runtime in production. To solve this, you can use a technique known as shadowing. Loom provides the `include` configuration which you can use like you are currently using `modImplementation` (in addition to `modImplementation`).

You probably should use `implementation` rather than `modImplementation` for that dependency since it doesn't need to be remapped. In general, only mod dependencies need `modImplementation`, and Java libraries used outside of the Minecraft/Fabric ecosystem can use `implementation`.

In total, your buildscript would look something like:

```gradle
dependencies {
    // ...
    implementation("org.tukaani:xz:1.10")
    include("org.tukaani:xz:1.10")
    // ...
}
```

[The Fabric Loom wiki page](https://wiki.fabricmc.net/documentation:fabric_loom) provides a few other examples of `include` and `implementation` that you can reference.

**Reply by IAmREGE** [reply]
> When you are running your mod in development, your dependencies will be available at runtime, but these dependencies are not available at runtime in production. To solve this, you can use a technique known as shadowing. Loom provides the `include` configuration which you can use like you are currently using `modImplementation` (in addition to `modImplementation`).
> 
> You probably should use `implementation` rather than `modImplementation` for that dependency since it doesn't need to be remapped. In general, only mod dependencies need `modImplementation`, and Java libraries used outside of the Minecraft/Fabric ecosystem can use `implementation`.
> 
> In total, your buildscript would look something like:
> 
> ```groovy-gradle
> dependencies {
>     // ...
>     implementation("org.tukaani:xz:1.10")
>     include("org.tukaani:xz:1.10")
>     // ...
> }
> ```
> 
> [The Fabric Loom wiki page](https://wiki.fabricmc.net/documentation:fabric_loom) provides a few other examples of `include` and `implementation` that you can reference.

Thanks. Will this cause conflict if another mod have also been done `include("org.tukaani:xz:1.10")`?

**Reply by haykam821** [reply]
No, two mods can include the same dependency. One issue that may arise is two mods including two versions of the same dependency, which may break if the two versions of the dependency have breaking changes and are not compatible with each other.

**Reply by sakura-ryoko** [reply]
My new favorite Java plugin is called ShadowJar for standalone projects.  I get the feeling that either fabric loader or Minecraft itself uses this.
